scripts/check-ci-job-skippability.js

```diff
@@ -0,0 +1,107 @@
+#!/usr/bin/env node
+
+/*
+
+This script takes two parameters or three environment variables:
+
+  {CI job name|$CIRCLE_JOB} {SHA1...SHA1|$CIRCLE_MERGE_BASE...$CIRCLE_SHA1}
+
+The flow of this script:
+
+  - List of files changed in PR: `git diff --name-only start...end`
+  - Extract list into array.
+  - Remove items from array if they match:
+    - Global ignore list.
+    - Other platform ignore lists.
+  - If array still contains items, do not authorize skipping CI job.
+  - Else if array is empty, skip CI job.
+
+*/
+
+const execSync = require('child_process').execSync;
+const _ = require('lodash');
+const minimatch = require("minimatch");
+
+console.step = _.partial(console.log, '\n\033[1m\033[36m*', _, '\033[0m');
+
+// BUILD PARAMETERS
+const CI_JOB_NAME = process.argv[2] || process.env.CIRCLE_JOB;
+if (!CI_JOB_NAME) { console.error('You must specify the name of a CI job as a parameter or environment variable.'); process.exit(1); }
+
+const COMMIT_RANGE = process.argv[3] || `${process.env.CIRCLE_MERGE_BASE}...${process.env.CIRCLE_SHA1}`;
+if (!COMMIT_RANGE || COMMIT_RANGE.includes('undefined')) { console.error('You must specify a range of commits to check for changed files in as a parameter or environment variable.'); process.exit(1); }
+
+// FILE SKIPPING PATTERNS
+const ALWAYS_SKIPPABLE_FILES = [
+  '**/CHANGELOG.md',
+  '**/README.md',
+  '**/CONTRIBUTING.md',
+  '**/DEVELOPING.md',
+  '**/INSTALL.md',
+  '**/*.podspec'
+]
+
+const ANDROID_FILES = [
+  'platform/android/**'
+]
+
+const IOS_FILES = [
+  'platform/ios/**',
+  'platform/darwin/**'
+]
+
+const MACOS_FILES = [
+  'platform/macos/**',
+  'platform/darwin/**'
+]
+
+const OTHER_PLATFORMS = [
+  'platform/glfw/**',
+  'platform/linux/**',
+  'platform/node/**',
+  'platform/qt/**',
+  'benchmark/**',
+  'test/**'
+]
+
+function skippableFilePatternsForCIJob(job) {
+  if (job.startsWith('ios')) {
+      return _.concat(ALWAYS_SKIPPABLE_FILES, ANDROID_FILES, MACOS_FILES, OTHER_PLATFORMS);
```

Comment:
Including macOS here means that `platform/darwin` will be skippable for iOS builds, which shouldn’t happen. Will need to think about the best way to deal with shared folders.

Reply:
Would it be easier to flip the logic so it checks for paths that should trigger a step (rather than not-triggering)?

Reply:
I think conceptually that might be easier, but I’m afraid that we’d miss some path we didn’t consider (or added/renamed later) and then the jobs would all keep quickly “passing” when they shouldn’t.

In general, I’d like to err on the side of over-running jobs and keeping this as a sort of “opt-in” to having your job/platform be skippable.

Reply:
Covered the `darwin` problem in https://github.com/mapbox/mapbox-gl-native/pull/15305/commits/a3675e09ed580ea305d8f73d52109bd9fdbf765c.